### ISSUE TITLE
Restoring SavedModel in //tensorflow/c:c_api_test fails 

### ISSUE THREAD

**Issue by namrata-ibm** [tensorflow/tensorflow#11290]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04 
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version**: 'v1.2.1-0-gb4957ff', '1.2.1'
- **Python version**: 2.7.12
- **Bazel version**: 0.4.5
- **CUDA/cuDNN version**: No GPU
- **GPU model and memory**: No GPU
- **Exact command to reproduce**: bazel test //tensorflow/c:c_api_test

### The problem:
While executing c_api_test from c module on a big endian machine, it fails while restoring SavedModel(tensorflow/cc/saved_model/testdata/half_plus_two/00000123). There is a warning displayed `Reading a bundle with different endianness from the reader`. 

Similarly, other tests from cc, java module which read this SavedModel from testdata also fail due to endianness mismatch.

What would be appropriate way to handle this failure on big endian?
Is there a way to convert the above SavedModel to Big Endian while reading it in the tests?

### Test Logs:
```
[ RUN      ] CAPI.SavedModel
2017-06-29 09:04:33.999662: I tensorflow/cc/saved_model/loader.cc:226] Loading SavedModel from: <HOME>/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/c/c_api_test.runfiles/org_tensorflow/tensorflow/cc/saved_model/testdata/half_plus_two/00000123
2017-06-29 09:04:34.039733: I tensorflow/cc/saved_model/loader.cc:145] Restoring SavedModel bundle.
2017-06-29 09:04:34.079006: W tensorflow/core/framework/op_kernel.cc:1158] Unimplemented: Reading a bundle with different endianness from the reader
2017-06-29 09:04:34.079311: W tensorflow/core/framework/op_kernel.cc:1158] Unimplemented: Reading a bundle with different endianness from the reader
2017-06-29 09:04:34.079363: W tensorflow/core/framework/op_kernel.cc:1158] Unimplemented: Reading a bundle with different endianness from the reader
2017-06-29 09:04:34.079708: I tensorflow/cc/saved_model/loader.cc:274] Loading SavedModel: fail. Took 136931 microseconds.
tensorflow/c/c_api_test.cc:1198: Failure
      Expected: TF_OK
      Which is: 0
To be equal to: TF_GetCode(s)
      Which is: 12
Reading a bundle with different endianness from the reader
         [[Node: save/RestoreV2 = RestoreV2[_output_shapes=[[]], dtypes=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/cpu:0"](_arg_save/Const_0_1, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]
```

**Comment by cy89** [comment]
There's a suggestion of a workaround for Endianness conversion in #2619. Also some discussion of how to set and handle the IS_LITTLE_ENDIAN flag.  
I think we're in the unfortunate situation of not being able to test big-endian code. We're happy to accept patches that add big-endian support.

**Comment by Nayana-ibm** [comment]
@cy89 Thanks for your reply. I have checked the flag used to detect underlying platform.  
The flag `kLittleEndian `declared [here ](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/core/platform/cpu_info.h#L29) can be used to check for big endian However, the `c_api_test`  fails on the SavedModel from test data which is in raw format and we are not sure about an appropriate way to convert this raw data suitable for big endian.

**Comment by Nayana-ibm** [comment]
Further analysis on this test shows that there is a check for endianness.  In case of BIG endian, test fails with "`Unimplemented: Reading a bundle with different endianness from the reader`". 
I tried to bypass this check and allowed the execution further for this test however it gives wrong output. 
The tensor bundle write/read seems to be working for big endian so now suspecting that the test data for big endian systems can be uploaded separately. But not sure how to generate the test data. 
Can you please confirm this understanding is correct? Also, could you let us know the steps which will generate test data. 
@aselle @poxvoculi Could you please provide your inputs on this?

**Comment by poxvoculi** [comment]
@asimshankar: Do you know who's in a position to fix c_api_test.cc?

**Comment by asimshankar** [comment]
I think the larger question raised by this is what our position should be regarding serialization formats. Regenerating the testdata would make this test pass, but we'd still have issues where say a SavedModel produced by a little endian machine will be unreadable on a big endian machine and vice-versa. Similarly, any tensors sent over RPCs across between machines with different architectures would likely fail.

A more principled fix would be something that perhaps includes the encoding format in the serialization and translates if necessary when deserializing. Something like that, I believe, would be the right way to go about ensuring TensorFlow is big-endian friendly.

That said, for specifically this test, you could regenerate a saved model in big-endian format in the testdata directory by running https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/saved_model/saved_model_half_plus_two.py , committing the output and changing `c_api_test.cc` to use one path or the other based on the architecture it's running on.

Unassigning myself as both changes outlined above is not something that I, or anyone else on the TensorFlow team (I think), will be getting to in the near future. But we'd be happy to accept PRs.

**Comment by poxvoculi** [comment]
@asimshankar: Model serialization is outside my area of expertise, but I saw this related issue last week: #11908  Why is model save/restore sensitive to endianness?  Are we not saving in protobuf format?

**Comment by asimshankar** [comment]
@poxvoculi : I'm not entirely sure about the history behind this as it predates my involvement in the project. Perhaps @zffchen78 has some historical knowledge?

But looking at the code, when we write out a tensor to disk (see [`tensor_bundle.cc`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/util/tensor_bundle/tensor_bundle.cc#L138)), it seems we're dumping the raw contents of the `TensorBuffer`, which will be in the endianness of the architecture it was on. Even when we dump to a proto ([`Tensor::AsProtoTensorContent`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/framework/tensor.cc#L727)), the [memory buffer is being used as-is](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/framework/tensor.cc#L122)). And in neither do we make a note of the endianness of the host when writing it out.

In other words, it doesn't look like we use the endian-safe routines in `coding.cc` for tensor content.

**Comment by Nayana-ibm** [comment]
@asimshankar @poxvoculi Thank you for your inputs. 

We could use the script [saved_model_half_plus_two.py](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/examples/saved_model/saved_model_half_plus_two.py) (referred in previous comment) to regenerate a saved model testdata in big-endian format (specifically s390x).  With the changes in source code` test/BUILD` files, now below tests are passing:
```
//tensorflow/c:c_api_test
//tensorflow/cc/saved_model:loader_test
//tensorflow/java:SavedModelBundleTest
//tensorflow/contrib/session_bundle:bundle_shim_py_test
//tensorflow/contrib/session_bundle:bundle_shim_test
```

On further execution of test modules like` tensorflow/contrib, tensorflow/tools` I found that there are few tests which are failing with same reason. i.e. testdata in little endian format. 
```
//tensorflow/python/feature_column:feature_column_test
//tensorflow/tools/tfprof/internal:tfprof_show_test
//tensorflow/tools/tfprof/internal:tfprof_stats_test
//tensorflow/tools/tfprof/internal:tfprof_tensor_test
//tensorflow/contrib/session_bundle:session_bundle_py_test
//tensorflow/contrib/session_bundle:session_bundle_test
```
We could also generate partial[ testdata for session bundle](https://github.com/tensorflow/tensorflow/tree/v1.2.1/tensorflow/contrib/session_bundle/testdata/half_plus_two/00000123) by executing script https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/contrib/session_bundle/example/export_half_plus_two.py used by session_bundle tests. 

Now,  we need to generate testdata used by below modules as well: 
1. `tensorflow/session_bundle/testdata/half_plus_two_ckpt_v2`
https://github.com/tensorflow/tensorflow/tree/v1.2.1/tensorflow/contrib/session_bundle/testdata/half_plus_two_ckpt_v2/00000123
2. `tensorflow/python/feature_column `
https://github.com/tensorflow/tensorflow/tree/v1.2.1/tensorflow/python/feature_column/testdata
3. ` tensorflow/tools/tfprof/internal/testdata`
https://github.com/tensorflow/tensorflow/tree/v1.2.1/tensorflow/tools/tfprof/internal/testdata


I tried to search for scripts/commands that could generate this testdata however didn't get much information. Am I missing something here?
 Could you please provide us pointers?

**Comment by poxvoculi** [comment]
@lilao can you provide guidance on session_bundle/testdata?

@ispirmustafa can you provide guidance on feature_column_test?

@panyx0718 can you advise on tprof/internal/testdata?

**Comment by panyx0718** [comment]
tprof/internal/testdata is a tiny model in python/profiler/model_analyzer_test.py. However, there is no script to automatically regenerate the data. And regenerating the data would require change of test results (due to time/memory dynamics).

The tests don't affect save/restore model. So, it should be safe to disable in your environment.

**Comment by ispirmustafa** [comment]
Same here, there is no script for feature_column_test and you can disable feature_column tests for your environment.

**Comment by asimshankar** [comment]
While we can skip tests/update testdata, this does mean that the serialized tensors will only be usable on the architecture that created it, which is a sad situation to be in.

@namrata-ibm : Would you have the bandwidth to contribute a more principled fix where we could record the endinaness of the creator in the checkpoint and do the conversion if necessary when we load the checkpoint?

**Comment by namrata-ibm** [comment]
@asimshankar, agree with you that this needs a better fix. However right now some other functionality needs attention on BE. Focusing on that first to make v1.2.1 available for BE and then consider this.

For now, skipping these tests for BE and accordingly raising a PR.

@asimshankar , @poxvoculi , @panyx0718 , @ispirmustafa  Thank you for your guidance.

**Comment by Nayana-ibm** [comment]
Looks liked fixed by #16540.
This issue can be closed

**Comment by Masquerade0097** [comment]
I'm still facing an error in loading a model on s390x from a protobuf file is still there. I think #41652 is related to his issue.

**CC** @Nayana-ibm @asimshankar
